plutus-emulator/src/Wallet/Emulator/Servant.hs

```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TupleSections         #-}
+{-# LANGUAGE TypeApplications      #-}
+module Wallet.Emulator.Servant where
+
+import           Conduit
+import           Control.Concurrent.STM       (atomically)
+import           Control.Concurrent.STM.TVar
+import           Control.Lens                 hiding (index)
+import           Control.Monad.Except
+import           Control.Monad.IO.Class       (MonadIO (..))
+import           Control.Monad.Reader
+import           Control.Monad.State          hiding (state)
+import           Control.Monad.Trans.Resource ()
+import           Data.List                    (partition)
+import           Data.Maybe                   (isNothing)
+import           Data.Traversable             (for)
+import           Servant                      (ServantErr (..))
+
+import           Ledger                       (Block, Blockchain, Slot (..), Tx (..), unspentOutputs)
+import qualified Ledger.Index                 as Index
+import qualified Ledger.Interval              as Interval
+
+-- | A pool of transactions which have yet to be validated.
+type TxPool = [Tx]
+
+data ServerState = ServerState {
+    _chainNewestFirst :: Blockchain, -- ^ The current chain, with the newest transactions first in the list.
+    _txPool           :: TxPool, -- ^ The pool of pending transactions.
+    _index            :: Index.UtxoIndex -- ^ The UTxO index, used for validation.
+} deriving (Show)
+
+makeLenses ''ServerState
+
+type MonadServer m = (MonadReader (TVar ServerState) m, MonadError ServantErr m, MonadIO m)
+
+publishTx :: MonadServer m => Tx -> m ()
+publishTx tx = do
+    tState <- ask
+    void . liftIO . atomically $ do
+        oldState <- readTVar tState
+        let newState = over txPool (tx :) oldState
+        writeTVar tState newState
+
+currentSlot :: MonadServer m => m Integer
+currentSlot = do
+    tState <- ask
+    let viewChain = view chainNewestFirst
+    liftIO . atomically $ fromIntegral . length . viewChain <$> readTVar tState
+
+txnStream :: MonadIO m => Integer -> TVar ServerState -> ConduitT () Block m ()
+txnStream slot state = loop slot
+    where
+        loop :: MonadIO m => Integer -> ConduitT () Block m ()
```

Comment:
I *think* you'd be able to do this over some not-necessarily-IO monad if you do what I suggested above. I'm not too familiar with conduit, though - maybe the notification stuff doesn't make sense unless you're in an asynchronous setting?